### PR TITLE
Agent to teardown sessions on signals

### LINKED ISSUE #1476: systemctl stop rust-iml-agent doesn't call daemon teardown
**Describe the bug**
When iml-agent stops, it does not call `teardown()` for the deamon plugins.

**To Reproduce**
1. `systemctl stop rust-iml-agent`
2. expect `teardown()` to be called and actions performed

### PR DESCRIPTION
**Pull request by ip1981**
Closes #1476.

```
jul 15 15:24:10 c1.local iml-agent-daemon[663]: Jul 15 15:24:10.810 DEBUG iml_agent::http_comms::crypto_client: resp headers: {"server": "nginx/1.16.1", "date": "Wed, 15 Jul 2020 15:24:10 GMT", "content-length": "0", "connection": "keep-alive"}
jul 15 15:24:13 c1.local systemd[1]: Stopping Rust IML Agent Daemon...
jul 15 15:24:13 c1.local iml-agent-daemon[663]: Jul 15 15:24:13.191  INFO iml_agent_daemon: Terminating...
jul 15 15:24:13 c1.local iml-agent-daemon[663]: Jul 15 15:24:13.191  INFO iml_agent::http_comms::session: Terminating all sessions
jul 15 15:24:13 c1.local iml-agent-daemon[663]: Jul 15 15:24:13.191  INFO iml_agent::http_comms::session: Terminating session PluginName("stats")/Id("40cf2205-aaee-44ef-9601-f6a38b306851")
jul 15 15:24:13 c1.local iml-agent-daemon[663]: Jul 15 15:24:13.192  INFO iml_agent::http_comms::session: Terminating session PluginName("ntp")/Id("928ba107-9443-4df2-9e86-da141bd958f2")
jul 15 15:24:13 c1.local iml-agent-daemon[663]: Jul 15 15:24:13.192  INFO iml_agent::http_comms::session: Terminating session PluginName("postoffice")/Id("8162b9b7-3fda-47a2-bc55-9fa3b12c860b")
jul 15 15:24:13 c1.local iml-agent-daemon[663]: Jul 15 15:24:13.192 DEBUG iml_agent::daemon_plugins::postoffice: Ending Inotify Listen for /etc/iml/postman.conf
jul 15 15:24:13 c1.local iml-agent-daemon[663]: Jul 15 15:24:13.192  INFO iml_agent::http_comms::session: Terminating session PluginName("device")/Id("13a704ce-f772-4caa-9233-649f27a1c85a")
jul 15 15:24:13 c1.local iml-agent-daemon[663]: Jul 15 15:24:13.193  INFO iml_agent::http_comms::session: Terminating session PluginName("ostpool")/Id("370515a3-0986-4b04-8b2a-db1dda484d3e")
jul 15 15:24:13 c1.local iml-agent-daemon[663]: Jul 15 15:24:13.193  INFO iml_agent::http_comms::session: Terminating session PluginName("action_runner")/Id("d64d353c-30ed-45c3-bb4e-a7163d580783")
jul 15 15:24:13 c1.local systemd[1]: Stopped Rust IML Agent Daemon.
jul 15 15:24:13 c1.local systemd[1]: Started Rust IML Agent Daemon.
jul 15 15:24:13 c1.local iml-agent-daemon[1828]: Jul 15 15:24:13.280  INFO iml_agent_daemon: Starting Rust agent_daemon
jul 15 15:24:13 c1.local iml-agent-daemon[1828]: unable to write 'random state'
jul 15 15:24:13 c1.local iml-agent-daemon[1828]: Jul 15 15:24:13.312  INFO iml_agent::daemon_plugins::daemon_plugin: Loaded the following DaemonPlugins:
```

Signed-off-by: Igor Pashev <pashev.igor@gmail.com>

---
This change is [Reviewable](https://reviewable.io/reviews/whamcloud/integrated-manager-for-lustre/2077)

### CODE REVIEW

File: iml-agent/src/main.rs [hunk 1]
```diff
@@ -31,15 +35,33 @@ async fn main() -> Result<()> {
...
+        poller::create_poller(agent_client, sessions.clone()),
+        poller_reg,
+    ));
+
+    let mut sigterm = signal(SignalKind::terminate()).expect("Could not listen to SIGTERM");
+    let mut sigint = signal(SignalKind::interrupt()).expect("Could not listen to SIGINT");
+    futures::select! {
+        () = sigterm.recv().fuse().map(drop) => (),
+        () = sigint.recv().fuse().map(drop) => ()
+    }
```
Comment (ip1981) [inline iml-agent/src/main.rs:59]: Probably we could make a utility function:

```rust
async fn main () {
  // spawn
  signals.await;
  // clean up
}
```

File: iml-agent/src/main.rs [hunk 2]
```diff
@@ -31,15 +35,33 @@ async fn main() -> Result<()> {
...
+
+    let (poller, poller_reg) = AbortHandle::new_pair();
+    tokio::spawn(Abortable::new(
+        poller::create_poller(agent_client, sessions.clone()),
+        poller_reg,
+    ));
+
+    let mut sigterm = signal(SignalKind::terminate()).expect("Could not listen to SIGTERM");
+    let mut sigint = signal(SignalKind::interrupt()).expect("Could not listen to SIGINT");
+    futures::select! {
```
Comment (jgrund) [inline iml-agent/src/main.rs:56]: Non-macro way to do this:

`future::select(sigterm.recv().boxed(), sigint.recv().boxed()).await;`

**Comment by jgrund** [comment]
Has a conflict on latest master

File: iml-agent/src/main.rs (hunk 2 above)
Comment (ip1981) [reply iml-agent/src/main.rs]: Better :)

**Pull request merged**